extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/PanacheQuery.java

```diff
@@ -37,9 +38,24 @@
      * retrieved from the database.
      *
      * @return a new query with the same state as the previous one (params, page, range, lockMode, hints, ...).
+     * @see #project(Function, String...)
      */
     public <T> PanacheQuery<T> project(Class<T> type);
 
+    /**
+     * Generates a SELECT clause with the specified properties.
+     * <p>
+     * The row mapper function is used to build a DTO object for each row of the result. A result row
+     * is represented by {@code Object[]}.
+     * 
+     * @param <T>
+     * @param rowMapper Maps a result row to a DTO object.
+     * @param properties
+     * @return a new query with the same state as the previous one (params, page, range, lockMode, hints, ...).
+     * @see #project(Class)
+     */
+    public <T> PanacheQuery<T> project(Function<Object[], T> rowMapper, String... properties);
```

Comment:
I would argue here, as well as in the kotlin form, to reverse the arguments.  Take a `List<String>` instead so devs can put the more complicated lambda at the end.  Or better yet on a new lilne.  It would be far more readable that way.  As it is now, the field names tend to get lost in the lambda noise.

```java
        List<Person> namesAndSurnames = Person.find("order by height desc")
                .project(Arrays.asList("name", "surname"), 
                    result -> {
                        Person p = new Person();
                        p.name = result[0].toString();
                        p.surname = result[1].toString();
                        return p;
                    }).stream().collect(Collectors.toList());
```

Reply:
Well, that was my original proposal but others thought that we should use varargs... @FroMage @loicmathieu WDYT?

Reply:
My problem with the varargs (beside the ordering) is that it makes it easy to have an empty set with out noticing.  There must be at least one argument which can't be enforced with varargs.  Yes, one could pass `List.of()` and get the same empty set but that's a much more conscious choice and less prone to accidentally fat fingering the invocation.

But the readability is the ultimate big win for me.

Reply:
I'll let @FroMage give his advise on this one, he's the one that ask to reverse the parameters and use varags but @evanchooly made a point, if the mapper is complex (and it will be if people map to a POJO) better have it at the end for readibility. So this seems to be a compromise of redeability vs writability ;)

Reply:
I prefer varargs because that _is_ the idiomatic way to do it in Java, even though I agree the order is inconvenient. Having to use `Arrays.asList` for _every_ call is a code smell: it means we're not using the language right.

As for requiring at least one argument, the idiomatic (albeit lame) way to do this in Java is:

```java
void foo(String one);
void foo(String one, String... more);
```

I do agree that we can reverse the order in Kotlin if that's idiomatic.